app/services/account_builder.rb

```diff
@@ -0,0 +1,185 @@
+# Builds a subclassed Account object given params.
+#
+# This class supports extensions from engines via subclassing `AccountBuilder`
+# object that are named `#{AccountSubclass}Builder` (i.e. NufsAccountBuilder).
+# See `self.for()` for additional detail.
+#
+# Example usage:
+#   factory = AccountBuilder.for("NufsAccount")
+#   account = factory.new(...).build
+#
+class AccountBuilder
+
+  attr_reader :account, :account_type, :current_user, :facility, :owner_user,
+    :params, :params_key
+
+  # Initialize the instance variables.
+  #
+  # Possible `options` include:
+  #   - account_type - subclassed `Account` class name required for new/build
+  #   - account - subclassed `Account` object required for update
+  #   - facility - a facility or facility null object
+  #   - current_user - required
+  #   - owner_user - passed only for new/build account objects
+  #   - params - required
+  #
+  def initialize(options = {})
+    @account_type = valid_account_type!(options[:account], options[:account_type], options[:facility])
+    @facility = options[:facility]
+    @owner_user = options[:owner_user]
+    @current_user = options[:current_user]
+    @params_key = Account.config.account_type_to_param(@account_type)
+    @params = parse_params(options[:account], options[:params], @params_key)
+    @account = options[:account] || new_account(@account_type)
+  end
+
+  # Factory method that returns a subclassed `AccountBuilder` if one exists for
+  # the given account_type; otherwise returns `AccountBuilder` class (self).
+  # Be careful that `Object.const_get()` doesn't provide false-negatives for
+  # lazy-loaded builder classes. The `account_type` argument accepts an
+  # `Account` class object, class name string/symbol, or underscored class name
+  # string/symbol.
+  #
+  # Example usage:
+  #   AccountBuilder.for("NufsAccount") => NufsAccountBuilder
+  #   AccountBuilder.for("WithoutBuilderAccount") => AccountBuilder
+  #
+  def self.for(account_type)
+    klass = "#{account_type}_builder".classify
+    Object.const_defined?(klass) ? klass.constantize : self
+  end
+
+  # Builds and returns a new account subclassed `Account` object.
+  # Used when creating an account.
+  def build
+    @account = assign_params(@account, @params, @params_key)
```

Comment:
To expand on our earlier slack chat...  The instance variables do not need to be passed around.  I've tried both approaches in the past and have started to prefer the more explicit approach.  Although I'm not strongly opinionated on this matter and could be talked into the less-explicit approach.  But a few pros do the explicit approach are:

* As a general rule of thumb, I like to test private methods.  Sometimes for TDD.  But also because if a class's public API breaks, the tests let us know immediately why the API is broken.  
* I've found it's easier/cleaner to unit test a method that accepts args and returns something; rather than setting up instance variables and checking them afterwards.  
* Passing the ivars as arguments is more explicit.  So it's more clear what goes into and comes out of a method.  And helps avoid unexpected/accidental modification of non-returned values.
* As a minor bonus side effect, avoiding instance variables in the private method decouples the methods from the class so it's easier to refactor or paste/copy.
* As far as I understand, we don't take performance hit or do extra garbage collection since ruby is a `pass-reference-by-value` language.  Although I may misunderstand how the memory management works in ruby.

```
Book = Struct.new("Book", :title)
book = Book.new(title: "foo")
book.object_id 
# => 70129693244980

book.title = "bar"
book.object_id 
# => 70129693244980

def change_title(book)
  puts "before change in method: #{book.object_id}"
  book.title = "changed"
  puts "after change in method: #{book.object_id}"
  book
end

book = change_title(book)
# => before change in method: 70129673327740
# => after change in method: 70129673327740

book.object_id
# => 70129673327740
```

I'll be in the office tomorrow, we should talk in person.  And I'll execute whatever you think is best.  :)